symbolic-debuginfo/src/bcsymbolmap.rs

```diff
@@ -0,0 +1,248 @@
+//! this case the actual symbol names are replaced with `__hidden#[0-9]+_` in the binary and
+//! dSYMs.  These can then be looked up in the BCSymbolMap to get the original symbol name
+//! back.
+//!
+//! See [`MachObject::load_symbolmap`](crate::macho::MachObject::load_symbolmap) for an
+//! example of how to use this.
+
+use std::error::Error;
+use std::fmt;
+use std::iter::FusedIterator;
+
+use thiserror::Error;
+
+use symbolic_common::DebugId;
+
+use crate::macho::SWIFT_HIDDEN_PREFIX;
+
+const BC_SYMBOL_MAP_HEADER: &str = "BCSymbolMap Version: 2.0";
+
+/// The error type for handling a [`BCSymbolMap`].
+#[derive(Debug, Error)]
+#[error("{kind}")]
+pub struct BCSymbolMapError {
+    kind: BCSymbolMapErrorKind,
+    #[source]
+    source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+
+/// Error kind for [`BCSymbolMapError`].
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum BCSymbolMapErrorKind {
+    /// The BCSymbolMap header does not match a supported version.
+    ///
+    /// It could be entirely missing, or only be an unknown version or otherwise corrupted.
+    InvalidHeader,
+    /// The bitcode symbol map did contain invalid UTF-8.
+    InvalidUtf8,
+}
+
+impl fmt::Display for BCSymbolMapErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::InvalidHeader => write!(f, "no valid BCSymbolMap header was found"),
+            Self::InvalidUtf8 => write!(f, "BCSymbolmap is not valid UTF-8"),
+        }
+    }
+}
+
+/// An in-memory representation of the Apple bitcode symbol map.
+///
+/// This is an auxiliary file, not an object file.  It can be used to provide de-obfuscated
+/// symbol names to a [`MachObject`] object using its
+/// [`load_symbolmap`](crate::macho::MachObject::load_symbolmap) method.
+///
+/// It is common for bitcode builds to obfuscate the names in the object file's symbol table
+/// so that even the DWARF files do not have the actual symbol names.  In this case the
+/// build process will create a `.bcsymbolmap` file which maps the obfuscated symbol names
+/// back to the original ones.  This structure can parse these files and allows providing
+/// this information to the [`MachObject`] so that it has the original symbol names instead
+/// of `__hidden#NNN_` ones.
+///
+/// [`MachObject`]: crate::macho::MachObject
+#[derive(Clone, Debug)]
+pub struct BCSymbolMap<'d> {
+    id: DebugId, // TODO
+    names: Vec<&'d str>,
+}
+
+impl From<BCSymbolMapErrorKind> for BCSymbolMapError {
+    fn from(source: BCSymbolMapErrorKind) -> Self {
+        Self {
+            kind: source,
+            source: None,
+        }
+    }
+}
+
+impl<'d> BCSymbolMap<'d> {
+    /// Tests whether the buffer could contain a [`BCSymbolMap`].
+    pub fn test(bytes: &[u8]) -> bool {
+        let mut pattern = BC_SYMBOL_MAP_HEADER.as_bytes();
+        if pattern.len() > bytes.len() {
+            pattern = &pattern[..bytes.len()];
+        }
```

Comment:
This means that a truncated file would pass this check. Is this really what we want? Or is this a limitation that `test` does get called with a small slice of the file only?

Reply:
Good point, actually all you want to do is `bytes.starts_with(pattern)`, right?

Reply:
Yes, the assumption is that this only gets called with a small slice.  This is kind of mimicking what `ObjectLike::test` does, this test is sufficient to decide between "is this an object? (`Archive::test`)", "is this a BCSymbolMap (`BCSymbolMap::test`)" or "is this a PList? (`PList::test`)".  So in a way the API assumption here is very weak and test is only useful in relation to the other test functions.
On the other hand to call `::parse()`, which is the only thing that actually gives you a full answer to this question, you need to read all the data from disk.  So this is useful as a quick check if it's even worth reading the entire file from disk.  Again this only makes sense if a BCSymbolMap is regularly larger than the filesystem block size, but it seems that is the case.
Anyway, TL;DR is that this is symmetric with `ObjectLike` and makes sense when using it in some of the workflows we do.